### PR TITLE
Add tests for `Object.groupBy` and `Map.groupBy` with string items

### LINKED ISSUE #4036: Add test case for `Object.groupBy` / `Map.groupBy` with strings?
`Object.groupBy` and `Map.groupBy` take an iterable, such as an array, as their first argument. This can be any iterable, so they can also accept a string.

```js
const str = "abcd";
const grouped = Object.groupBy(str, (char) => char < 'c' ? "x" : "y");
```

However, it was discovered that in some implementations, a TypeError was thrown when the first argument was not of object type, so this was not working[^1].

What do you think about adding test cases to test262 to cover this?

[^1]: https://bugs.webkit.org/show_bug.cgi?id=271524

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Fixes #4036

### CODE REVIEW

File: test/built-ins/Map/groupBy/string.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+esid: sec-map.groupby
+description: Map.groupBy works for string items
+info: |
+  Map.groupBy ( items, callbackfn )
+  ...
+includes: [compareArray.js]
+features: [array-grouping, Map]
+---*/
+
+const string = 'abcd';
```
Comment (ljharb) [inline test/built-ins/Map/groupBy/string.js:14]: let's please add a surrogate pair into the test strings. i typically use 💩 but pick your favorite :-)

File: test/built-ins/Map/groupBy/string.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+const map = Map.groupBy(string, function (char) {
+  return char < 'c' ? 'before' : 'after';
+});
+
+assert.compareArray(Array.from(map.keys()), ['before', 'after']);
+assert.compareArray(map.get('before'), ['a', 'b']);
+assert.compareArray(map.get('odd'), ['c', 'd']);
```
Comment (ljharb) [inline test/built-ins/Map/groupBy/string.js:22]:
```suggestion
assert.compareArray(map.get('after'), ['c', 'd']);
```

File: test/built-ins/Object/groupBy/string.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+const string = 'abcd';
+
+const obj = Object.groupBy(string, function (char) {
+  return char < 'c' ? 'before' : 'after';
+});
+
+assert.compareArray(Object.keys(obj), ['before', 'after']);
+assert.compareArray(obj['before'], ['a', 'b']);
+assert.compareArray(obj['after'], ['c', 'd']);
```
Comment (ljharb) [inline test/built-ins/Object/groupBy/string.js:22]:
```suggestion
assert.compareArray(obj.before, ['a', 'b']);
assert.compareArray(obj.after, ['c', 'd']);
```

File: test/built-ins/Map/groupBy/string.js (hunk 1 above)
Comment (sosukesuzuki) [reply test/built-ins/Map/groupBy/string.js]: I chose four emojis. [961e633](https://github.com/tc39/test262/pull/4038/commits/961e63382699256ec7f072a9255c9bf6fc619ed9)

File: test/built-ins/Map/groupBy/string.js [hunk 4]
```diff
@@ -11,12 +11,12 @@ includes: [compareArray.js]
 features: [array-grouping, Map]
 ---*/
 
-const string = 'abcd';
+const string = '🥰💩🙏😈';
```
Comment (ljharb) [inline test/built-ins/Map/groupBy/string.js:14]: oh i didn't mean _just_ surrogate pairs :-) this is fine, but i was expecting a mix of code units and code points.

**Review by ljharb — APPROVED** [review]

**Review by ptomato — APPROVED** [review]
Not much to add to what Jordan said.

For me this is also fine to merge as is, but as long as we're on the topic of grouping strings with weird edge cases: if you felt like doing extra work, I think it might help to have coverage for ASCII strings (to help debuggability in case an implementation gets single characters right but surrogate pairs wrong) and grapheme clusters (like 👨‍👦 being split into 👨, ZWJ, and 👦.)

Thanks for contributing!

**Comment by ptomato** [comment]
@sosukesuzuki Works for me. I'll merge this now and copy my comment into #4036 for whoever picks it up.

**Pull request merged**